### PR TITLE
HPCC-26390 Add support for packages to archives

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Also need to merge https://github.com/hpcc-systems/LN/pull/993 to match changes

**Comment by AttilaVamos** [comment]
@ghalliday There are tons of errors like this:
```
      1. Test: HeadingExample.ecl
      eclcc returns with:
        Error (1001): Command return code non-zero.
         retCode: 2
        /home/centos/smoketest/PR-15331/HPCC-Platform/testing/regress/ecl/HeadingExample.ecl(1,0): error C0003: assert(name) failed - file: hqlexpr.cpp, line 1334
        1 error, 0 warning

```

File: ecl/eclcc/eclcc.cpp [hunk 1]
```diff
@@ -2157,9 +2190,9 @@ bool EclCC::processFiles()
     }
     else if (inputFiles.ordinality() == 0)
     {
-        assertex(optQueryRepositoryReference);
+        assertex(optQueryMainAttribute);
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp:2193]: This assertex is not necessary as the tests conditions & logic in lines 2144-2150 means that this cannot ever be false.

File: ecl/hql/hqlrepository.cpp [hunk 2]
```diff
@@ -544,7 +557,32 @@ void EclRepositoryManager::addMapping(const char * url, const char * path)
     repos.append(*new EclRepositoryMapping(repo, version, path));
 }
 
-IEclRepository * EclRepositoryManager::resolveDependentRepository(IIdAtom * name, const char * defaultUrl, bool requireSHA)
+void EclRepositoryManager::processArchive(IPropertyTree * archiveTree)
+{
+    IArrayOf<IEclRepository> savedSources;        // also includes -D options
+    savedSources.swapWith(allSources);
+
+    Owned<IErrorReceiver> errs = createThrowingErrorReceiver();
```
Comment (shamser) [inline ecl/hql/hqlrepository.cpp:565]: I don't see 'errs' being used anywhere

File: ecl/hql/hqlexpr.cpp [hunk 3]
```diff
@@ -1341,6 +1366,24 @@ IPropertyTree * queryEnsureArchiveModule(IPropertyTree * archive, const char * n
     return module;
 }
 
+void ensureArchiveImport(IPropertyTree * archive, const char * package, IIdAtom * id, IHqlScope * scope)
```
Comment (shamser) [inline ecl/hql/hqlexpr.cpp:1369]: What's this function doing?  |May be a comment line to explain.

File: ecl/eclcc/eclcc.cpp [hunk 4]
```diff
@@ -2768,8 +2801,20 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
         else if (iter.matchOption(optMonitorInterval, "--monitorinterval"))
         {
         }
-        else if (iter.matchOption(optQueryRepositoryReference, "-main"))
+        else if (iter.matchOption(tempArg, "-main"))
         {
+            const char * arg = tempArg;
+            const char * at = strchr(arg, '@');
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp:2807]: The command's help/usage could be updated.

**Review by shamser — COMMENTED** [review]
@ghalliday

**Review by shamser — APPROVED** [review]

**Pull request merged**